multiple variable length argument is not checked in generics method

<!-- Please make sure to run `v up` before reporting any issues as it may have already been fixed.
     It's also advisable to update all relevant modules using `v outdated` and `v install` -->

<!-- You can use `v doctor` to fill up the next fields -->
**V version: 0.3.1 7fb2718** 
**OS: ArchLinux** 

<!-- Please include the standalone minimal reproducible code.
If not, it may be closed by the administrators. -->
**What did you do?**
```v
fn foo(i ...int){}

fn bar<T>(i ...int){
	foo(i)
}
```

**What did you expect to see?**
Error message like
```v
fn foo(i ...int){}

fn bar(i ...int){
	foo(i)
}
```
```
bug.v:4:2: error: to pass `i` (int) to `foo` (which accepts type `...int`), use `...i`
    2 | 
    3 | fn bar(i ...int){
    4 |     foo(i)
      |     ~~~~~~
    5 | }
    6 |
```

**What did you see instead?**
No error.

hmm.. complex real program produces build error
```
/tmp/v_1000/first.5791279522746268595.tmp.c:47439: error: cannot convert 'struct array' to 'unsigned char *'
...
then I thought this no checking causes it..
I should make real minimal reproduce..

Ok, this is real minimal reproducible example.

```v
pub struct Client{}

pub fn (mut cl Client) req<T>(data ...string) {}

pub struct Product{}

pub fn (mut c Client) products_list(options ...string) {
	c.req<Product>(options) // (...options) works
}

fn main(){
	mut sc := Client{}
	sc.products_list()
}
```
```
==================
/tmp/v_1000/bug.6871599712269147543.tmp.c:12107: error: cannot convert 'struct array' to 'unsigned char *'
...
==================
```

--- a/vlib/v/checker/fn.v
+++ b/vlib/v/checker/fn.v
@@ -941,8 +941,8 @@ pub fn (mut c Checker) fn_call(mut node ast.CallExpr, mut continue_check &bool) 
 			typ := c.expr(call_arg.expr)
 			if i == node.args.len - 1 {
 				if c.table.sym(typ).kind == .array && call_arg.expr !is ast.ArrayDecompose
-					&& c.table.sym(expected_type).kind != .array && !param.typ.has_flag(.generic)
-					&& expected_type != typ {
+					&& c.table.sym(expected_type).kind !in [.sum_type, .interface_]
+					&& !param.typ.has_flag(.generic) && expected_type != typ {
 					styp := c.table.type_to_str(typ)
 					elem_styp := c.table.type_to_str(expected_type)
 					c.error('to pass `$call_arg.expr` ($styp) to `$func.name` (which accepts type `...$elem_styp`), use `...$call_arg.expr`',
@@ -1551,7 +1551,7 @@ pub fn (mut c Checker) method_call(mut node ast.CallExpr) ast.Type {
 				typ := c.expr(arg.expr)
 				if i == node.args.len - 1 {
 					if c.table.sym(typ).kind == .array && arg.expr !is ast.ArrayDecompose
-						&& c.table.sym(expected_type).kind != .array
+						&& c.table.sym(expected_type).kind !in [.sum_type, .interface_]
 						&& !param.typ.has_flag(.generic) && expected_type != typ {
 						styp := c.table.type_to_str(typ)
 						elem_styp := c.table.type_to_str(expected_type)

--- a/vlib/v/checker/fn.v
+++ b/vlib/v/checker/fn.v
@@ -1533,6 +1533,41 @@ pub fn (mut c Checker) method_call(mut node ast.CallExpr) ast.Type {
 			} else {
 				method.params[i + 1]
 			}
+
+			if method.is_variadic && arg.expr is ast.ArrayDecompose {
+				if i > method.params.len - 2 {
+					c.error('too many arguments in call to `$method.name`', node.pos)
+				}
+			}
+			if method.is_variadic && i >= method.params.len - 2 {
+				param_sym := c.table.sym(param.typ)
+				mut expected_type := param.typ
+				if param_sym.kind == .array {
+					info := param_sym.array_info()
+					expected_type = info.elem_type
+					c.expected_type = expected_type
+				}
+				typ := c.expr(arg.expr)
+				if i == node.args.len - 1 {
+					if c.table.sym(typ).kind == .array && arg.expr !is ast.ArrayDecompose
+						&& !param.typ.has_flag(.generic) && expected_type != typ {
+						styp := c.table.type_to_str(typ)
+						elem_styp := c.table.type_to_str(expected_type)
+						c.error('to pass `$arg.expr` ($styp) to `$method.name` (which accepts type `...$elem_styp`), use `...$arg.expr`',
+							node.pos)
+					} else if arg.expr is ast.ArrayDecompose
+						&& c.table.sym(expected_type).kind == .sum_type
+						&& expected_type.idx() != typ.idx() {
+						expected_type_str := c.table.type_to_str(expected_type)
+						got_type_str := c.table.type_to_str(typ)
+						c.error('cannot use `...$got_type_str` as `...$expected_type_str` in argument ${
+							i + 1} to `$method_name`', arg.pos)
+					}
+				}
+			} else {
+				c.expected_type = param.typ
+			}
+
 			param_is_mut = param_is_mut || param.is_mut
 			param_share := param.typ.share()
 			if param_share == .shared_t && (c.locked_names.len > 0 || c.rlocked_names.len > 0) {

--- a/vlib/v/checker/fn.v
+++ b/vlib/v/checker/fn.v
@@ -941,7 +941,8 @@ pub fn (mut c Checker) fn_call(mut node ast.CallExpr, mut continue_check &bool) 
 			typ := c.expr(call_arg.expr)
 			if i == node.args.len - 1 {
 				if c.table.sym(typ).kind == .array && call_arg.expr !is ast.ArrayDecompose
-					&& !param.typ.has_flag(.generic) && expected_type != typ {
+					&& c.table.sym(expected_type).kind != .array && !param.typ.has_flag(.generic)
+					&& expected_type != typ {
 					styp := c.table.type_to_str(typ)
 					elem_styp := c.table.type_to_str(expected_type)
 					c.error('to pass `$call_arg.expr` ($styp) to `$func.name` (which accepts type `...$elem_styp`), use `...$call_arg.expr`',
@@ -1550,6 +1551,7 @@ pub fn (mut c Checker) method_call(mut node ast.CallExpr) ast.Type {
 				typ := c.expr(arg.expr)
 				if i == node.args.len - 1 {
 					if c.table.sym(typ).kind == .array && arg.expr !is ast.ArrayDecompose
+						&& c.table.sym(expected_type).kind != .array
 						&& !param.typ.has_flag(.generic) && expected_type != typ {
 						styp := c.table.type_to_str(typ)
 						elem_styp := c.table.type_to_str(expected_type)

--- a/vlib/v/checker/tests/generics_method_called_variadic_arg_mismatch.out
+++ b/vlib/v/checker/tests/generics_method_called_variadic_arg_mismatch.out
@@ -1,0 +1,7 @@
+vlib/v/checker/tests/generics_method_called_variadic_arg_mismatch.vv:8:4: error: to pass `options` (string) to `req` (which accepts type `...string`), use `...options`
+    6 |
+    7 | fn (c Client) products_list(options ...string) {
+    8 |     c.req<Product>(options) // (...options) works
+      |       ~~~~~~~~~~~~~~~~~~~~~
+    9 | }
+   10 |

--- a/vlib/v/checker/tests/generics_method_called_variadic_arg_mismatch.vv
+++ b/vlib/v/checker/tests/generics_method_called_variadic_arg_mismatch.vv
@@ -1,0 +1,14 @@
+struct Client {}
+
+fn (cl Client) req<T>(data ...string) {}
+
+struct Product {}
+
+fn (c Client) products_list(options ...string) {
+	c.req<Product>(options) // (...options) works
+}
+
+fn main() {
+	mut sc := Client{}
+	sc.products_list()
+}

